contracts/lens/OpinionatedOracle.sol

```diff
@@ -0,0 +1,102 @@
+        Weak, // 1 minute
+        Medium, // 30 minutes
+        Strong // 1 day
+    }
+
+    uint32[4] quotePeriods = [0, 60 seconds, 30 minutes, 1 days];
+
+    /// @param _factory Address of the UniswapV3 Factory used to look up token pools
+    constructor(address _factory) {
+        factory = IUniswapV3Factory(_factory);
+    }
+
+    /// @notice Returns a quote for a given token amount according to a desired price manipulation resistance
+    /// @notice across a specified list of fee tiers
+    /// @param baseToken Address of the input token for the quote
+    /// @param quoteToken Address of the output token for the quote
+    /// @param baseTokenAmount The amount of token to be converted
+    /// @param resistance The desired resistance against price manipulation: Dangerous, Weak, Medium, Strong
+    /// @param feeTiers An array of fee tiers to gather prices from for the quote
+    /// @return quoteTokenAmount The amount of quoteToken received for baseTokenAmount of baseToken
+    function quote(
+        address baseToken,
+        address quoteToken,
+        uint128 baseTokenAmount,
+        ManipulationResistance resistance,
+        uint24[] memory feeTiers
+    ) public view returns (uint256 quoteTokenAmount) {
+        require(feeTiers.length > 0, 'FT');
+
+        uint32 quotePeriod = quotePeriods[uint256(resistance)];
+        address pool;
+        int24 meanWeightedTick;
+
+        // If the quote period is more than 0, we can calculate the mean tick over an interval and weight it by
+        // each pool's mean liquidity over the same period
+        if (quotePeriod > 0) {
+            WeightedOracleLibrary.PeriodObservation[] memory poolObservations =
+                new WeightedOracleLibrary.PeriodObservation[](feeTiers.length);
+
+            for (uint256 i = 0; i < feeTiers.length; i++) {
+                pool = factory.getPool(baseToken, quoteToken, feeTiers[i]);
+                require(pool != address(0), 'NP');
+                poolObservations[i] = WeightedOracleLibrary.consult(pool, quotePeriod);
+            }
+
+            meanWeightedTick = WeightedOracleLibrary.getArithmeticMeanTickWeightedByLiquidity(poolObservations);
+        } else {
+            // This is a 'Dangerous' quote request where instead of an average price over an interval,
+            // a spot price is fetched and used.
+            int256 totalWeightedTicks;
+            uint128 totalLiquidity;
+            int24 tick;
+            uint128 liquidity;
+
+            for (uint256 i = 0; i < feeTiers.length; i++) {
+                pool = factory.getPool(baseToken, quoteToken, feeTiers[i]);
+                require(pool != address(0), 'NP');
+                (tick, liquidity) = OracleLibrary.getBlockStartingTickAndLiquidity(pool);
+                totalWeightedTicks += int256(liquidity) * tick;
+                totalLiquidity += liquidity;
+            }
+
+            meanWeightedTick = int24(totalWeightedTicks / totalLiquidity);
```

Comment:
the way this number is rounded should be consistent (towards negative infinity iirc)

Reply:
you could maybe construct weighted observations here and use the same method above

but this is pretty jank anyway, the nice thing about harmonic mean liquidity is that it is [extremely biased](https://www.wolframalpha.com/input/?i=harmonic+mean+of+%7B1%2C+2%5E128%2C+2%5E128%2C+2%5E128%2C+2%5E128%2C+2%5E128%2C+2%5E128%7D) towards lower values, but if you can get the block ending liquidity to be high on a pool you can bias the result a lot. then again i guess that's why we'll call this one dangerous.